map/extrapolation/extrapolator.cpp

```diff
@@ -0,0 +1,162 @@
+#include "map/extrapolation/extrapolator.hpp"
+
+#include "geometry/distance_on_sphere.hpp"
+
+#include "platform/platform.hpp"
+
+#include "base/logging.hpp"
+
+#include <chrono>
+#include <memory>
+
+namespace
+{
+uint64_t constexpr kMaxExtrapolationTimeMs = 1000;
+uint64_t constexpr kExtrapolationPeriodMs = 200;
+double constexpr kMaxExtrapolationSpeedMPerS = 120.0;
+
+double LinearExtrapolationOfOneParam(double param1, double param2, uint64_t timeBetweenPointsMs,
+                                     uint64_t timeAfterPoint2Ms)
+{
+  return (param2 - param1) * timeAfterPoint2Ms / timeBetweenPointsMs;
+}
+}  // namespace
+
+namespace position_extrapolator
+{
+using namespace std;
+
+location::GpsInfo LinearExtrapolation(location::GpsInfo const & point1, location::GpsInfo const & point2,
+                                      uint64_t timeAfterPoint2Ms)
+{
+  auto const timeBetweenPointsMs =
+      static_cast<uint64_t>((point2.m_timestamp - point1.m_timestamp) * 1000);
+
+  location::GpsInfo extrapolated = point2;
+
+  extrapolated.m_timestamp += timeAfterPoint2Ms;
+
+  extrapolated.m_longitude += LinearExtrapolationOfOneParam(
+      point1.m_longitude, point2.m_longitude, timeBetweenPointsMs, timeAfterPoint2Ms);
+
+  extrapolated.m_latitude += LinearExtrapolationOfOneParam(point1.m_latitude, point2.m_latitude,
+                                                           timeBetweenPointsMs, timeAfterPoint2Ms);
+
+  extrapolated.m_horizontalAccuracy +=
+      LinearExtrapolationOfOneParam(point1.m_horizontalAccuracy, point2.m_horizontalAccuracy,
+                                    timeBetweenPointsMs, timeAfterPoint2Ms);
+  extrapolated.m_altitude += LinearExtrapolationOfOneParam(point1.m_altitude, point2.m_altitude,
+                                                           timeBetweenPointsMs, timeAfterPoint2Ms);
+
+  if (point1.m_verticalAccuracy != -1 && point2.m_verticalAccuracy != -1)
+  {
+    extrapolated.m_verticalAccuracy +=
+        LinearExtrapolationOfOneParam(point1.m_verticalAccuracy, point2.m_verticalAccuracy,
+                                      timeBetweenPointsMs, timeAfterPoint2Ms);
+  }
+
+  if (point1.m_bearing != -1 && point2.m_bearing != -1)
```

Comment:
I'm not sure about exact comparison of doubles here. It seems that methods similar to GpsInfo::HasBearing() and GpsInfo::HasSpeed() must be used instead.

Reply:
Right. Done. And I added a method HasVerticalAccuracy()